clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java

```diff
@@ -0,0 +1,226 @@
+        }
+    }
+
+    private void setConfig() {
+        this.retryBackoffMs = this.config.getLong(ConsumerConfig.RETRY_BACKOFF_MS_CONFIG);
+        this.clientId = config.getString(CommonClientConfigs.CLIENT_ID_CONFIG);
+        this.heartbeatIntervalMs = config.getInt(ConsumerConfig.HEARTBEAT_INTERVAL_MS_CONFIG);
+    }
+
+    @Override
+    public void run() {
+        try {
+            log.debug("{} started", getClass());
+            while (running) {
+                try {
+                    runOnce();
+                } catch (WakeupException | InterruptException e) {
+                    log.debug("Exception thrown, background thread won't " +
+                            "terminate", e);
+                    // swallow
+                }
+            }
+        } catch (Throwable t) {
+            log.error("The background thread failed due to unexpected error",
+                    t);
+            if (t instanceof RuntimeException)
+                this.exception.set(Optional.of((RuntimeException) t));
+            else
+                this.exception.set(Optional.of(new RuntimeException(t)));
+        } finally {
+            close();
+            log.debug("{} closed", getClass());
+        }
+    }
+
+    /**
+     * Process event from a single poll
+     */
+    void runOnce() {
+        this.inflightEvent = maybePollEvent();
+        log.debug("processing application event: {}", this.inflightEvent);
+        if (this.inflightEvent.isPresent() && maybeConsumeInflightEvent(this.inflightEvent.get())) {
+            // clear inflight event upon successful consumption
+            this.inflightEvent = Optional.empty();
+        }
+
+        // if there are pending events to process, poll then continue without
+        // blocking.
+        if (!applicationEventQueue.isEmpty() || inflightEvent.isPresent()) {
+            networkClient.poll(time.timer(0));
+            return;
+        }
+        // if there are no events to process, poll until timeout. The timeout
+        // will be the minimum of the requestTimeoutMs, nextHeartBeatMs, and
+        // nextMetadataUpdate. See NetworkClient.poll impl.
+        networkClient.poll(time.timer(timeToNextHeartbeatMs(time.milliseconds())));
+    }
+
+    private long timeToNextHeartbeatMs(long nowMs) {
+        // TODO: implemented when heartbeat is added to the impl
+        return 100;
+    }
+
+    private Optional<ApplicationEvent> maybePollEvent() {
+        if (this.inflightEvent.isPresent() || this.applicationEventQueue.isEmpty()) {
+            return this.inflightEvent;
+        }
+        return Optional.ofNullable(this.applicationEventQueue.poll());
+    }
+
+    /**
+     * ApplicationEvent are consumed here.
+     * @param event an {@link ApplicationEvent}
+     * @return true when successfully consumed the event.
+     */
+    private boolean maybeConsumeInflightEvent(ApplicationEvent event) {
```

Comment:
What's the benefit of having an `EventType` in addition to the `ApplicationEvent`? The type is in the incoming `ApplicationEvent` which appears to be then used in the `switch` statement to cast to the desired `ApplicationEvent` type. Why not just have add an abstract `process` method on the `ApplicationEvent` that does whatever needs to be done?

Reply:
If we let the `ApplicationEvent` perform whatever logic is needed, that frees up the `DefaultBackgroundThread` from having to encapsulate all the logic of all the events that can happen. Keeping the `ApplicationEvent` and the `DefaultBackgroundThread` a little more loosely-coupled may allow easier unit testing 🤷‍♂️

Reply:
Then you'd also not need to introduce the `NOOP` bits into this code at all.

Reply:
great idea